lib/semantics/check-do.cc

```diff
@@ -647,20 +645,131 @@ void DoChecker::Leave(const parser::DoConstruct &x) {
+  return std::visit(
+      [&](const auto &x) { return GetConstructName(*x); }, construct);
+}
+
+template<typename A> static parser::CharBlock GetConstructPosition(const A &a) {
+  return std::get<0>(a.t).source;
+}
+
+static parser::CharBlock GetNodePosition(const ConstructNode &construct) {
+  return std::visit(
+      [&](const auto &x) { return GetConstructPosition(*x); }, construct);
+}
+
+void DoChecker::SayBadLeave(const char *stmtChecked,
+    const char *enclosingStmtType, const ConstructNode &construct) const {
+  context_
+      .Say(*context_.location(), "%s must not leave a %s statement"_err_en_US,
+          stmtChecked, enclosingStmtType)
+      .Attach(GetNodePosition(construct), "The construct that was left"_en_US);
+}
+
+static parser::DoConstruct const *MaybeGetDoConstruct(
+    const ConstructNode &construct) {
+  if (const auto doNode{std::get_if<const parser::DoConstruct *>(&construct)}) {
+    return *doNode;
+  }
+  return nullptr;
+}
+
+// Check that CYCLE and EXIT statements do not cause flow of control to
+// leave DO CONCURRENT, CRITICAL, or CHANGE TEAM constructs.
+void DoChecker::CheckForBadLeave(
+    const char *stmtName, const ConstructNode &construct) const {
+  // C1135 and C1167
+  if (const auto doConstructPtr{MaybeGetDoConstruct(construct)}) {
+    if (doConstructPtr->IsDoConcurrent()) {
+      SayBadLeave(stmtName, "DO CONCURRENT", construct);
+    }
+    return;
+  }
+  // C1135 and C1168
+  if (const auto criticalConstructPtr{
+          std::get_if<const parser::CriticalConstruct *>(&construct)}) {
+    SayBadLeave(stmtName, "CRITICAL", construct);
+    return;
+  }
+  if (const auto changeTeamConstructPtr{
+          std::get_if<const parser::ChangeTeamConstruct *>(&construct)}) {
+    SayBadLeave(stmtName, "CHANGE TEAM", construct);
+    return;
+  }
+}
+
+static bool ConstructIsDoConcurrent(const ConstructNode &construct) {
+  parser::DoConstruct const *doConstruct{MaybeGetDoConstruct(construct)};
+  return doConstruct && doConstruct->IsDoConcurrent();
+}
+
+static bool isExitStmt(const char *stmtType) {
+  return std::strncmp("EXIT", stmtType, 4) == 0;
+}
+
+static bool StmtMatchesConstruct(NamePtr stmtName, const char *stmtType,
+    NamePtr constructName, const ConstructNode &construct) {
+  bool inDoConstruct{MaybeGetDoConstruct(construct) != nullptr};
+  if (stmtName == nullptr) {
+    if (inDoConstruct) {
+      return true;  // Unlabeled statements match all DO constructs
+    } else {
+      return false;  // Unlabeled statements match no non-DO constructs
+    }
+  } else if (!constructName) {
+    return false;  // name on CYCLE/EXIT, but not on the construct
+  } else if (constructName->source == stmtName->source) {
+    if (isExitStmt(stmtType)) {
+      return true;  // EXIT name matches any construct name
+    } else {
+      if (inDoConstruct) {
+        return true;  // CYCLE name matches only DO construct name
+      }
+    }
+  }
+  return false;
```

Comment:
This return expression could be a lot simpler:
```
  if (stmtName == nullptr) {
    return inDoConstruct;
  } else if (constructName && constructName->source == stmtName->source) {
    return isExitStmt(stmtType) || inDoConstruct;
  } else {
    return false;
  }
```

Reply:
Yes.  I'll change it as suggested.